### ISSUE TITLE
COPOD Explainability on unseen data

### ISSUE THREAD

**Issue by thewall27** [yzhao062/pyod#308]
COPOD explain_outlier() function works only for a given data point within the dataset. Is there any approach for explainability on unseen data( data on which model has not been trained) ?

**Comment by psmgeelen** [comment]
I have a similar issue. I would like to 1) have a method by which I can predict or 'explain' the individual scores of each feature, and I would like to do that predictively too. Is there a way I can contribute that feature? I notice that the decision_functgion sums the results. A simple flag to not do that would already be awesome! 

```
    def decision_function(self, X):
        """Predict raw anomaly score of X using the fitted detector.
         For consistency, outliers are assigned with larger anomaly scores.
        Parameters
        ----------
        X : numpy array of shape (n_samples, n_features)
            The training input samples. Sparse matrices are accepted only
            if they are supported by the base estimator.
        Returns
        -------
        anomaly_scores : numpy array of shape (n_samples,)
            The anomaly score of the input samples.
        """
        # use multi-thread execution
        if self.n_jobs != 1:
            return self._decision_function_parallel(X)

        if hasattr(self, 'X_train'):
            original_size = X.shape[0]
            X = np.concatenate((self.X_train, X), axis=0)

        self.U_l = pd.DataFrame(-1 * np.log(np.apply_along_axis(ecdf, 0, X)))
        self.U_r = pd.DataFrame(-1 * np.log(np.apply_along_axis(ecdf, 0, -X)))
        skewness = np.sign(skew(X, axis=0))
        self.U_skew = self.U_l * -1 * np.sign(
            skewness - 1) + self.U_r * np.sign(skewness + 1)
        self.O = np.maximum(self.U_skew, np.add(self.U_l, self.U_r) / 2)
        if hasattr(self, 'X_train'):
            decision_scores_ = self.O.sum(axis=1).to_numpy()[-original_size:]
        else:
            decision_scores_ = self.O.sum(axis=1).to_numpy()
        return decision_scores_.ravel()
```
Am I correct in the statement that the data for all the individual scores is stored in the attribute 'O'? This would already provide me with access to the individual scores of the trained data at least..

**Comment by yzhao062** [comment]
> COPOD explain_outlier() function works only for a given data point within the dataset. Is there any approach for explainability on unseen data( data on which model has not been trained) ?

The current functionality does not support it...but you could combine your test dataset with train to achieve this, indirectly.

**Comment by yzhao062** [comment]
> I have a similar issue. I would like to 1) have a method by which I can predict or 'explain' the individual scores of each feature, and I would like to do that predictively too. Is there a way I can contribute that feature? I notice that the decision_functgion sums the results. A simple flag to not do that would already be awesome!
> 
> ```
>     def decision_function(self, X):
>         """Predict raw anomaly score of X using the fitted detector.
>          For consistency, outliers are assigned with larger anomaly scores.
>         Parameters
>         ----------
>         X : numpy array of shape (n_samples, n_features)
>             The training input samples. Sparse matrices are accepted only
>             if they are supported by the base estimator.
>         Returns
>         -------
>         anomaly_scores : numpy array of shape (n_samples,)
>             The anomaly score of the input samples.
>         """
>         # use multi-thread execution
>         if self.n_jobs != 1:
>             return self._decision_function_parallel(X)
> 
>         if hasattr(self, 'X_train'):
>             original_size = X.shape[0]
>             X = np.concatenate((self.X_train, X), axis=0)
> 
>         self.U_l = pd.DataFrame(-1 * np.log(np.apply_along_axis(ecdf, 0, X)))
>         self.U_r = pd.DataFrame(-1 * np.log(np.apply_along_axis(ecdf, 0, -X)))
>         skewness = np.sign(skew(X, axis=0))
>         self.U_skew = self.U_l * -1 * np.sign(
>             skewness - 1) + self.U_r * np.sign(skewness + 1)
>         self.O = np.maximum(self.U_skew, np.add(self.U_l, self.U_r) / 2)
>         if hasattr(self, 'X_train'):
>             decision_scores_ = self.O.sum(axis=1).to_numpy()[-original_size:]
>         else:
>             decision_scores_ = self.O.sum(axis=1).to_numpy()
>         return decision_scores_.ravel()
> ```
> 
> Am I correct in the statement that the data for all the individual scores is stored in the attribute 'O'? This would already provide me with access to the individual scores of the trained data at least..

That is true. self.O is a numpy matrix with shape (n_samples, n_features), so it stores each individual scores.

**Comment by psmgeelen** [comment]
@yzhao062 would this be a feature that we could add? I dont mind doing the leg-work for it :)

**Comment by psmgeelen** [comment]
I prepped a pull-request with the appropiate changes. Please check whether I did a good job at making the request, as it is my first time. https://github.com/yzhao062/pyod/pull/309

**Comment by psmgeelen** [comment]
@thewall27 , my approach should give you all the functionality that you would need over the `decision_function` function, by adding the flag `scores_per_feature`. What this be acceptable to you?

**Comment by thewall27** [comment]
So greater the score of a feature, more chance that it is contributing for an observation to be an anomaly. Right? @psmgeelen

**Comment by psmgeelen** [comment]
@thewall27 , exactly! And this can be done with new data and old data. I still want to look at the predict_proba method that is stored in the base.py, but this contains a lot of data-formatting that I am not 100% comfortable yet with. The simple proba method is based on a MinMax scaler however, so outside of the library it could be wrapped as follows:

```
### Train model and scaler first
model = COPOD().fit(X)
predictions = COPOD.decision_boundaries(X, scores_per_feature=True)
scaler = MinMax().fit(predictions)

### How to get the proba's
probas =  scaler.transform(model.decision_boundaries(X, scores_per_feature=True)).clip(0,1)
```
Clip is a kewl method from Numpy (I found out 10 mins ago) that makes sure that values do not exceed (in this case) zero or 1.

**Comment by psmgeelen** [comment]
So it should be 

```
### Train model and scaler first
model = COPOD().fit(X)
predictions = model.decision_boundaries(X, scores_per_feature=True)
scaler = MinMax().fit(predictions)

### How to get the proba's
probas =  scaler.transform(model.decision_boundaries(X, scores_per_feature=True)).clip(0,1)
```
But for some reason I cannot change prior comment..

**Comment by cherepanovic** [comment]
>COPOD explain_outlier() function works only for a given data point within the dataset. Is there any approach for explainability on unseen data( data on which model has not been trained) ?

it is, in general, the case for all models because of all scaling/normalization functions... what to do if the plan is to make predictions on live data which surely will not always fit into recalculated scales of training set? 

thanks!

**Comment by psmgeelen** [comment]
@cherepanovic, so there are a couple of tricks that can be applied, but lets start at the basics. By definition one could argue that data that is out of range of the training data, is an anomaly. Also note how all scalers can be used to rescale all other data. This is never a problem in itself, as 'trained' scalers are simple arithmetic. Of course there are algorithms that need the data in a particular fashion, and 'test' can end up being out of range of that algorithm. This is seldomly the case though because the estimation processes are often based on 'distance' between 'common' and less 'common' data-points. For this reason it tends to measurable too. Please take a deep dive in what anomaly detection is; this library is an excellent example what common approaches are used. Its really worthwhile!
